### PR TITLE
Rewrite of the pokemon identification logic

### PR DESCRIPTION
**Pull request by nahojjjen**
This rewrite adds the ability to identify pokemon by evolution cost + candy name. The full chain of logic used to determine the pokemon in this implementation is:
1. check if the nickname perfectly matches a pokemon
2. check if closest candyname + evolution cost perfectly matches a pokemon
3. check if there's a cached result for the scanned pokemon name
4. get the pokemon with the closest name within the evolution line guessed from the candy

Known issue: Does not handle multiple evolutions such as eeveelutions

### CODE REVIEW

File: app/src/main/java/com/kamron/pogoiv/OCRHelper.java [hunk 1]
```diff
@@ -143,6 +144,36 @@ private double getPokemonLevelFromImg(Bitmap pokemonImage, int trainerLevel) {
...
+    public int getPokemonEvolutionCostFromImg(Bitmap pokemonImage) {
+        Bitmap evolutionCostImage =
+                Bitmap.createBitmap(pokemonImage, (int) (widthPixels * 0.625), (int) (heightPixels * 0.86),
+                        (int) (widthPixels * 0.2), (int) (heightPixels * 0.05));
+        String hash = "candyCost" + hashBitmap(evolutionCostImage);
+        String stringCacheEvoCandyCost = ocrCache.get(hash);
+        int result = -1;
+        if (stringCacheEvoCandyCost == null) { //if no cached result
+            //the dark color used for text in pogo is 76,112,114
+            evolutionCostImage = replaceColors(evolutionCostImage, 76, 112, 114, Color.WHITE, 40, false);
```
Comment (sarav) [inline app/src/main/java/com/kamron/pogoiv/OCRHelper.java]: If I remember the discussion right, you had to call replaceColors twice, but I don't see that here. How did you get around it?

File: app/src/main/java/com/kamron/pogoiv/OCRHelper.java [hunk 2]
```diff
@@ -143,6 +144,36 @@ private double getPokemonLevelFromImg(Bitmap pokemonImage, int trainerLevel) {
...
+            tesseract.setImage(evolutionCostImage);
+            stringCacheEvoCandyCost = tesseract.getUTF8Text().replace("S", "5").replace("s", "5");
+            try {
+                result = Integer.parseInt(stringCacheEvoCandyCost);
+            } catch (NumberFormatException e) {
+                stringCacheEvoCandyCost = "-1";
+            }
+            ocrCache.put(hash, stringCacheEvoCandyCost);
+        }
+        System.out.println("asdasdasd");
```
Comment (sarav) [inline app/src/main/java/com/kamron/pogoiv/OCRHelper.java]: Remove?

File: app/src/main/java/com/kamron/pogoiv/OCRHelper.java [hunk 3]
```diff
@@ -169,7 +169,6 @@ public int getPokemonEvolutionCostFromImg(Bitmap pokemonImage) {
             }
             ocrCache.put(hash, stringCacheEvoCandyCost);
         }
-        System.out.println("asdasdasd");
```
Comment (sarav) [inline app/src/main/java/com/kamron/pogoiv/OCRHelper.java]: It's fine to have these commits in your local branch, but when you think it's good enough for a PR, please squash these "local mistakes".

File: app/src/main/java/com/kamron/pogoiv/OCRHelper.java [hunk 4]
```diff
@@ -149,28 +149,76 @@ private double getPokemonLevelFromImg(Bitmap pokemonImage, int trainerLevel) {
...
+        if (stringCacheEvoCandyCost != null) return Integer.parseInt(stringCacheEvoCandyCost);
+
+        //clean the image
+        //the dark color used for text in pogo is approximately rgb 76,112,114 if you can afford evo
+        //and the red color is rgb 255 95 100 when you cant afford the evolution
+        Bitmap evolutionCostImageCanAfford = replaceColors(evolutionCostImage, 80, 110, 110, Color.WHITE, 45, false);
+        Bitmap evolutionCostImageCannotAfford = replaceColors(evolutionCostImage, 255, 95, 100, Color.WHITE, 30,
+                false);
+
+        boolean affordIsBlank = isOnlyWhite(evolutionCostImageCanAfford);
```
Comment (sarav) [inline app/src/main/java/com/kamron/pogoiv/OCRHelper.java]: Can we just keep track of this as part of replacing pixels in replaceColors? That avoids
1. Copying out a few pixels first
2. Accessing that memory again

**Comment by sarav** [comment]
Ok, this PR has waaay too many merges, merge conflict fixes and "local error" commits (mistakes you made while coding). Please rebase to master, clean up the commits and push the PR again. This is too messy to just merge into master. I'll wait till then to do further review -- because I'll have to review again from the start anyway to make sure no errors were introduced.

**Comment by sarav** [comment]
Btw, not sure about your comfort level with git. Let me know if you'd like me to help out with the rebases, squashes or any other git help (DM me on discord if you have questions about git commands).

**Comment by nahojjjen** [comment]
@sarav are you happy enough if the pr is merged with the squash & merge option? 
asdasdasd

I feel like any more complicated effort spent making the commit history cleaner would be a waste of time.

**Comment by sarav** [comment]
@nahojjjen I didn't see the discord chat discussion. I'll definitely try to test it out and report it (assuming I can merge it with the current master which has the battery saving feature).

Making a clean commit history is definitely not a waste of time -- from years of experience with git on a large project. I can go into the details of why it matters for large and small projects if you are interested (or maybe I'll comment later anyway). But I don't want to alienate you either by taking too strong a position about git commits. Your work/features are very valuable.

I don't use the git GUI (not have I used the github merging side of it), so I'm not exactly sure if it'll do the right thing. If it'll do what I think it does, then IMHO it'll be the very bare minimum for a large feature PR to be merged.

But the squashed commit it'll be a huge pain in the ass to do code review (or for a new dev to go back and understand the changes), which I think we'll eventually have to do before merging a significant feature/change like this. Code review is the best way to catch bugs.

If you want, I can take your branch and try to clean it up (mainly remove all the weird merges, merge conflicts) and push it back and then you can sync it and push it back as your PR or to continue your work. I'll obviously preserve the authorship. Let me know if that's something you don't mind.

**Comment by nahojjjen** [comment]
@sarav If you have the time and motivation to clean up this pr, please do. I currently don't have a lot of free time because of uni, and I do not have the knowledge required to do any major changes with commits.

**Comment by sarav** [comment]
I'll get around to it during the weekends maybe. But in the unlikely case I don't, please go with the squash merge.

**Comment by nahojjjen** [comment]
Pull request closed until sarav fixes commits

**Comment by sarav** [comment]
Oh... you could have still left it open so people know it's here and needs to be tested. Ok, I'll get to it this weekend -- I'm been missing out on SOOO MUCH sleep due to working on GoIV the past week!

**Comment by DJCrapsody** [comment]
Can we reopen this PR so everyone is aware that its being worked on?

I'm not sure if multi evolutions like eevees are fixed yet, we can identify the evolved form the same way as the nidorans. Maybe discussion was there and I probably missed it.

**Comment by sarav** [comment]
@DJCrapsody I think he's expecting the discussion to continue in the PR I pushed.

File: app/src/main/java/com/kamron/pogoiv/logic/PokemonNameCorrector.java [hunk 5]
```diff
@@ -120,4 +173,117 @@ public PokeDist getPossiblePokemon(String poketext, String candytext) {
...
+
+        ArrayList<Pokemon> bestGuessEvolutionLine = getBestGuessForEvolutionLine(candytext);
+
+        Pokemon bestMatchPokemon = null;
+        int lowestDist = Integer.MAX_VALUE;
+        for (Pokemon trypoke : bestGuessEvolutionLine) {
+
+            int dist = trypoke.getDistanceCaseInsensitive(poketext);
+            if (dist < lowestDist) {
+                bestMatchPokemon = trypoke;
```
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/logic/PokemonNameCorrector.java]: So `lowestDist` wasn't updated here... just noticed while reworking #502.

**Pull request closed without merging**